rust/src/bwrap.rs

```diff
@@ -0,0 +1,481 @@
+    tempdir: Option<tempfile::TempDir>,
+}
+
+impl RoFilesMount {
+    /// Create a new rofiles-fuse mount point
+    fn new(rootfs: &openat::Dir, path: &str) -> Result<Self> {
+        let path = path.trim_start_matches('/');
+        let tempdir = tempfile::Builder::new()
+            .prefix("rpmostree-rofiles-fuse")
+            .tempdir()?;
+        let status = std::process::Command::new("rofiles-fuse")
+            .arg("--copyup")
+            .arg(path)
+            .arg(tempdir.path())
+            .current_dir(format!("/proc/self/fd/{}", rootfs.as_raw_fd()))
+            .status()?;
+        if !status.success() {
+            return Err(anyhow::anyhow!("{}", status));
+        }
+        Ok(Self {
+            tempdir: Some(tempdir),
+        })
+    }
+
+    /// Return the mount point path
+    fn path(&self) -> &Path {
+        // Safety: We only use the Option<T> here around drop handling, it should always
+        // be `Some` until `drop()` is called.
+        self.tempdir.as_ref().unwrap().path()
+    }
+}
+
+impl Drop for RoFilesMount {
+    fn drop(&mut self) {
+        let tempdir = if let Some(d) = self.tempdir.take() {
+            d
+        } else {
+            return;
+        };
+        // We need to unmount before letting the tempdir cleanup run.
+        let success = Command::new("fusermount")
+            .arg("-u")
+            .arg(tempdir.path())
+            .status()
+            .map_err(anyhow::Error::new)
+            .and_then(|status| -> Result<()> {
+                if !status.success() {
+                    Err(anyhow::anyhow!("{}", status))
+                } else {
+                    Ok(())
+                }
+            })
+            .err()
+            .map(|e| {
+                systemd::journal::print(4, &format!("{}", e));
+            })
+            .is_none();
+        if !success {
+            // If fusermount fails, then we cannot remove it; just leak it.
+            let _ = tempdir.into_path();
+        }
+    }
+}
+
+/// Helper wrapper that waits for a child and checks its exit status.
+/// Further if the wait is cancelled then the child is force killed.
+fn child_wait_check(
+    child: gio::Subprocess,
+    cancellable: Option<&gio::Cancellable>,
+) -> std::result::Result<(), glib::Error> {
+    match child.wait(cancellable) {
+        Ok(_) => {
+            let estatus = child.get_exit_status();
+            glib::spawn_check_exit_status(estatus)?;
+            Ok(())
+        }
+        Err(e) => {
+            child.force_exit();
+            Err(e)
+        }
+    }
+}
+
+impl Bubblewrap {
+    /// Create a new Bubblewrap instance
+    pub(crate) fn new(rootfs_fd: &openat::Dir) -> Result<Self> {
+        let rootfs_fd = rootfs_fd.sub_dir(".")?;
+
+        let lang = std::env::var_os("LANG");
+        let lang = lang.as_ref().map(|s| s.to_str()).flatten().unwrap_or("C");
+        let lang_var = format!("LANG={}", lang);
+        let lang_var = Path::new(&lang_var);
```

Comment:
Aside: I'm confused why the glib bindings for this don't use the more appropriate `OsStr` instead.